Add Auto Complete to PostgREST example (#310)
pgroonga/pgroonga.github.io: how-to/postgrest.md

        <script src="https://cdn.jsdelivr.net/npm/@tarekraafat/autocomplete.js@10.2.7/dist/autoComplete.min.js">
        </script>
        <script type="text/javascript">
            const searchForm = document.getElementById('searchForm');

            searchForm.addEventListener('submit', async (event) => {
                event.preventDefault();
                const query = document.getElementById('autoComplete').value;
                
                if (query.length === 0) {
                    return;
                }

                const params = new URLSearchParams({keywords: query});
                const response = await fetch(`http://localhost:3000/rpc/find_title?${params.ToString()}`);
                const result = await response.json();

                const preElement = document.getElementById('output');
                if (result.length > 0) {
                    const formattedValue = JSON.stringify(result, null, 2);
                    preElement.textContent = formattedValue;
                } else {
                    preElement.textContent = 'No results found';
                }
            });

            const dataSource = async (query) => {

>                 const source = await fetch(`http://localhost:3000/rpc/autocomplete?keyword=${query}`);

Reviewer: Could you use `URLSearchParams` here too?

Author: I just figure out that using `URLSearchParams` will break not only auto complete feature but search function too.

Author: My bad. I misunderstood how to use `URLSearchParams`.

            };

            const autoCompleteJS = new autoComplete({
                placeHolder: "Enter a keyword",
                data: {
                    src: dataSource
                },
                searchEngine: (query, record) => record,
                events: {
                    input: {
                        selection: (event) => {
                            const selection = event.detail.selection.value;
                            autoCompleteJS.input.value = selection;
                        }
                    }
                },
                submit: true
            });
        </script>
        <pre id="output"></pre>
    </div>
  </body>
</html>
```

### Run PostgREST as API backend

Run your PostgREST service using following command:

```sh
postgrest memo.conf
```

### Open html and Try Out

Open `index.html` with your browser. 

![PGroonga Auto Complete1](../images/postgrest/auto-complete1.png)

Type something and it will show the suggestions.

![PGroonga Auto Complete2](../images/postgrest/auto-complete2.png)

When you press `Search` button, it will performe keyword search on memos table title data.

> ![PGroonga Auto Complete3](../images/postgrest/auto-complete3.png)

Reviewer: ```suggestion
![PGroonga Auto Complete3](../images/postgrest/auto-complete3.png)

[auto-complete]: auto-complete.html
```

]
```

This seamless and user-friendly approach to searching through your data with keywords is both practical and exciting. It's a fantastic way to enhance your search capabilities, don't you think?

## Using Keyword Auto Complete

PGroonga has features to implement auto complete which is explained in [this how to section](https://pgroonga.github.io/how-to/auto-complete.html).

> Here we will explor how to implement this using PostgREST and just a simple html with JavaScript.

Reviewer: ```suggestion
Here we will explore how to implement this using PostgREST and just a simple HTML with JavaScript.
```

### Create Table for Auto Complete Feature

```sql
CREATE TABLE terms (
  term text,
  readings text[]
);

CREATE INDEX pgroonga_terms_prefix_search ON terms USING pgroonga
  (readings pgroonga_text_array_term_search_ops_v2);

CREATE INDEX pgroonga_terms_full_text_search ON terms USING pgroonga
  (term)
  WITH (tokenizer = 'TokenBigramSplitSymbolAlphaDigit');

INSERT INTO terms (term, readings) VALUES ('牛乳', ARRAY['ギュウニュウ', 'ミルク']);
INSERT INTO terms (term, readings) VALUES ('PostgreSQL', ARRAY['ポスグレ', 'posugure', 'postgres']);
INSERT INTO terms (term, readings) VALUES ('Groonga', ARRAY['guru-nga', 'グルンガ','ぐるんが','gurunga']);
INSERT INTO terms (term, readings) VALUES ('PGroonga', ARRAY['pi-ji-runga', 'ピージールンガ','ぴーじーるんが','pg','ピージー']);
```

### Set Up PostgREST Permission

```sql

> RANT SELECT ON terms TO web_user;

Reviewer: ```suggestion
GRANT SELECT ON terms TO web_user;
```

```

### Create Auto Complete End Point

```sql
CREATE OR REPLACE FUNCTION autocomplete(keyword text) RETURNS SETOF text AS $$
DECLARE
  result text[];
BEGIN
  IF keyword = '' THEN
    RETURN QUERY SELECT unnest(result);
  ELSE
    RETURN QUERY SELECT term FROM terms WHERE readings &^~ keyword;
  END IF;
END;
$$ LANGUAGE plpgsql;
```

> ### Create a html with JavaScript

Reviewer: ```suggestion
### Create a HTML with JavaScript
```

Create following HTML file:

```sh
vi index.html
```

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

>     <title>Groonga dictionary search</title>

Reviewer: Could you update the title?

Author: Changed to "PGroonga Auto Complete Search"

    <link rel="stylesheet"
          href="https://cdn.jsdelivr.net/npm/@tarekraafat/autocomplete.js@10.2.7/dist/css/autoComplete.min.css">
    <style>
        .center-container {
            display: flex;
            justify-content: center;
            align-items: center;
            min-height: 100vh;
        }
    </style>
  </head>
  <body>
    <div class="center-container">
        <form name="search" id="searchForm">
            <input type="search" size="60" maxlength="60" name="key" id="autoComplete">
        </form>

        <script src="https://cdn.jsdelivr.net/npm/@tarekraafat/autocomplete.js@10.2.7/dist/autoComplete.min.js">
        </script>
        <script type="text/javascript">
            const searchForm = document.getElementById('searchForm');

            searchForm.addEventListener('submit', async (event) => {
                event.preventDefault();
                const query = document.getElementById('autoComplete').value;
                

>                 if(query.length === 0) {

Reviewer: ```suggestion
                if (query.length === 0) {
```

                    return;
                }

>                 const response = await fetch(`http://localhost:3000/rpc/find_title?keywords=${query}`);

Reviewer: We need to escape `query`:

```suggestion
                const params = new URLSearchParams({keywords: query});
                const response = await fetch(`http://localhost:3000/rpc/find_title?${params.ToString()}`);
```

Author: I thought PostgREST will prevent SQL Injection from happening by design. https://postgrest.org/en/stable/references/configuration.html#db-prepared-statements

But it is nice to be extra careful on this😀

Reviewer: This is not for SQL injection.
It's just for passing correct HTTP query parameter. If `query` is ab&c=d`, URL will be `http://localhost:3000/rpc/find_title?keywords=ab&c=d`. In the URL, `keywords` is `ab` not `ab&c=d`.

                const result = await response.json();

                const preElement = document.getElementById('output');

>                 if(result.length > 0) {

Reviewer: ```suggestion
                if (result.length > 0) {
```

                    const formattedValue = JSON.stringify(result, null, 2);
                    preElement.textContent = formattedValue;
                } else {
                    preElement.textContent = 'No results found';
                }
            });

>             const displayItems = (completions) => { };
> 

Reviewer: ```suggestion
```

            const dataSource = async (query) => {
                const source = await fetch(`http://localhost:3000/rpc/autocomplete?keyword=${query}`);

>                 const data = await source.json();
>                 const completions = data;
>                 displayItems(completions);
>                 return completions;

Reviewer: ```suggestion
                return await source.json();
```

            };

            const autoCompleteJS = new autoComplete({
                placeHolder: "Enter a keyword",
                data: {
                    src: dataSource
                },
                searchEngine: (query, record) => record,
                events: {
                    input: {
                        selection: (event) => {
                            const selection = event.detail.selection.value;
                            autoCompleteJS.input.value = selection;

>                             dataSource(selection);

Reviewer: ```suggestion
```

                        }
                    }
                },
                submit: true
            });
        </script>

>         <div id="result" style="margin-top: 7em;"></div>

Reviewer: ```suggestion
```

        <pre id="output"></pre>
    </div>
  </body>
</html>
```

### Run PostgREST as API backend

Run your PostgREST service using following command:

```sh
postgrest memo.conf
```

### Open html and Try Out

Open `index.html` with your browser. 

> ![PGroonga Auto Complete1](../images/postgres/pgautocomplete1.png)

Reviewer: Could you fix a typo (`postgres`) in path?

```suggestion
![PGroonga Auto Complete1](../images/postgrest/pgautocomplete1.png)
```

Could you use `auto-complete1.png`, `auto-complete2.png` and so on naming?
(We don't need `pg` prefix because we only focus on PostgreSQL. We want to use `-` as a word separator.)

Author: Hai yorokonde!

]
```

This seamless and user-friendly approach to searching through your data with keywords is both practical and exciting. It's a fantastic way to enhance your search capabilities, don't you think?

## Using Keyword Auto Complete

> PGroonga has features to implement auto complete which is explained in [this how to section](https://pgroonga.github.io/how-to/auto-complete.html).

Reviewer: Could you use relative link to reuse the same link path for all languages? And could you use named link (`[label][link-id]` and `[link-id]: URL`) instead of inline link (`[label][URL]`) for easy to translate?

```suggestion
PGroonga has features to implement auto complete which is explained in [the auto complete how to section][auto-complete].
```

and put the following to the page footer:

```markdown
[auto-complete]: auto-complete.html
```
